packages/access-api/src/service/access-delegate.js

```diff
@@ -0,0 +1,107 @@
+import { createDelegationsStorage } from './delegations.js'
+
+/**
+ * access/delegate failure due to the 'with' resource not having
+ * enough storage capacity to store the delegation.
+ * https://github.com/web3-storage/specs/blob/7e662a2d9ada4e3fc22a7a68f84871bff0a5380c/w3-access.md?plain=1#L94
+ *
+ * Semantics inspired by https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/507
+ *
+ * @typedef {import('@web3-storage/capabilities/types').InsufficientStorage} InsufficientStorage
+ */
+
+/**
+ * @typedef {import('@web3-storage/capabilities/types').AccessDelegateSuccess} AccessDelegateSuccess
+ * @typedef {import('@web3-storage/capabilities/types').AccessDelegateFailure} AccessDelegateFailure
+ * @typedef {Ucanto.Result<AccessDelegateSuccess, AccessDelegateFailure>} AccessDelegateResult
+ */
+
+/**
+ * @param {object} ctx
+ * @param {import('../types/delegations').DelegationsStorage} ctx.delegations
+ * @param {HasStorageProvider} ctx.hasStorageProvider
+ */
+export function accessDelegateProvider(ctx) {
+  const handleInvocation = createAccessDelegateHandler(ctx)
+  return Server.provide(delegate, async ({ capability, invocation }) => {
+    return handleInvocation(
+      /** @type {Ucanto.Invocation<import('@web3-storage/capabilities/types').AccessDelegate>} */ (
+        invocation
+      )
+    )
+  })
+}
+
+/**
+ * @callback AccessDelegateHandler
+ * @param {Ucanto.Invocation<import('@web3-storage/capabilities/types').AccessDelegate>} invocation
+ * @returns {Promise<AccessDelegateResult>}
+ */
+
+/**
+ * @callback HasStorageProvider
+ * @param {Ucanto.DID<'key'>} did
+ * @returns {Promise<boolean>} whether the given resource has a storage provider
+ */
+
+/**
+ * @param {object} options
+ * @param {import('../types/delegations').DelegationsStorage} [options.delegations]
+ * @param {HasStorageProvider} [options.hasStorageProvider]
+ * @param {boolean} [options.allowServiceWithoutStorageProvider] - whether to allow service if the capability resource does not have a storage provider
+ * @returns {AccessDelegateHandler}
+ */
+export function createAccessDelegateHandler({
+  delegations = createDelegationsStorage(),
+  hasStorageProvider = async () => false,
+  allowServiceWithoutStorageProvider = false,
+} = {}) {
+  return async (invocation) => {
+    const capabability = invocation.capabilities[0]
+    if (
+      !allowServiceWithoutStorageProvider &&
+      !(await hasStorageProvider(capabability.with))
+    ) {
+      return {
+        name: 'InsufficientStorage',
+        message: `${capabability.with} has no storage provider`,
+        error: true,
+      }
+    }
+    const delegated = extractProvenDelegations(invocation)
+    await delegations.push(...delegated)
+    return {}
+  }
+}
+
+/**
+ * @param {Ucanto.Invocation<import('@web3-storage/capabilities/types').AccessDelegate>} invocation
+ * @returns {Iterable<Ucanto.Delegation<Ucanto.Capabilities>>}
+ */
+function* extractProvenDelegations({ proofs, capabilities }) {
```

Comment:
Maybe personal preference but I'd prefer collecting in an array to using a generator unless there's some kind of streaming benefit to be had.

Reply:
defining this as a generator function yields delegations as we enounter them, which affords for streaming them out.

You can trivially collect them into an array like:
```
const collected = Array.from(extractProvenDelegations(invocation))
```

but if I authored this logic as a non-generator, it wouldn't be as useful to anyone that wanted to stream or e.g. only extract the first N proven delegation.

Reply:
I don't think I saw a significant amount of work that the function was doing that would benefit from it being done incrementally.

When this function is being used, you immediately spread the results into a function call, which is not using the generator in a streaming fashion anyway.

If in the future you need to consume part of the stream, or it is blocking the event loop or you want to do some work between recieving each proof then sure, make it a generator. Right now though, this function reads to _me_ as though it does some heavy lifting or as a consumer I'm going to want/need to process the data incrementally, which doesn't appear to be the case.

Reply:
This is not blocking!